optimizely/decision/composite_service.go

```diff
@@ -18,38 +18,56 @@ package decision
 
 import (
 	"github.com/optimizely/go-sdk/optimizely/entities"
+	"github.com/optimizely/go-sdk/optimizely/notification"
 )
 
 // CompositeService is the entrypoint into the decision service. It provides out of the box decision making for Features and Experiments.
 type CompositeService struct {
 	experimentDecisionServices []ExperimentDecisionService
 	featureDecisionServices    []FeatureDecisionService
+	notificationCenter         notification.Center
 }
 
 // NewCompositeService returns a new instance of the DefeaultDecisionEngine
 func NewCompositeService() *CompositeService {
 	featureDecisionService := NewCompositeFeatureService()
 	return &CompositeService{
 		featureDecisionServices: []FeatureDecisionService{featureDecisionService},
+		notificationCenter:      notification.NewNotificationCenter(),
 	}
 }
 
 // GetFeatureDecision returns a decision for the given feature key
 func (s CompositeService) GetFeatureDecision(featureDecisionContext FeatureDecisionContext, userContext entities.UserContext) (FeatureDecision, error) {
 	var featureDecision FeatureDecision
-
+	var err error
 	// loop through the different features decision services until we get a decision
 	for _, decisionService := range s.featureDecisionServices {
-		featureDecision, err := decisionService.GetDecision(featureDecisionContext, userContext)
+		featureDecision, err = decisionService.GetDecision(featureDecisionContext, userContext)
 		if err != nil {
 			// @TODO: log error
 		}
 
 		if featureDecision.DecisionMade {
-			return featureDecision, err
+			break
 		}
 	}
 
 	// @TODO: add errors
-	return featureDecision, nil
+
+	// @TODO convert the decision to a notification.DecisionNotification
+	if s.notificationCenter != nil {
```

Comment:
Hmm good question. We don't always control the instantiation of the implementation. That is, a user can always just say `s := decision.CompositeService{}` and pass that into the client and not pass in a notification center instead of calling our `decision.CompositeService()` factory method. The question is then, how well do we guard against these possibilities since?